sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,788 @@
+      return false;
+    }
+
+    @Override
+    public BoundedReader<String> createReader(PipelineOptions options) throws IOException {
+      return new BoundedElasticsearchReader(this);
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public Coder<String> getDefaultOutputCoder() {
+      return StringUtf8Coder.of();
+    }
+
+    private JsonObject getStats(boolean shardLevel) throws IOException {
+      RestClient restClient = spec.getConnectionConfiguration().createClient();
+      HashMap<String, String> params = new HashMap<>();
+      if (shardLevel) {
+        params.put("level", "shards");
+      }
+      String endpoint = String.format("/%s/_stats", spec.getConnectionConfiguration().getIndex());
+      Response response;
+      try {
+        response = restClient.performRequest("GET", endpoint, params, new BasicHeader("", ""));
+      } finally {
+        restClient.close();
+      }
+      return parseResponse(response);
+    }
+  }
+
+  private static class BoundedElasticsearchReader extends BoundedSource.BoundedReader<String> {
+
+    private final BoundedElasticsearchSource source;
+
+    private RestClient restClient;
+    private String current;
+    private String scrollId;
+    private List<String> batch;
+
+    public BoundedElasticsearchReader(BoundedElasticsearchSource source) {
+      this.source = source;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      restClient = source.spec.getConnectionConfiguration().createClient();
+
+      String query = source.spec.getQuery();
+      if (query == null) {
+        query = "{ \"query\": { \"match_all\": {} } }";
+      }
+
+      Response response;
+      String endPoint =
+          String.format(
+              "/%s/%s/_search",
+              source.spec.getConnectionConfiguration().getIndex(),
+              source.spec.getConnectionConfiguration().getType());
+      Map<String, String> params = new HashMap<>();
+      params.put("scroll", source.spec.getScrollKeepalive());
+      params.put("size", String.valueOf(source.spec.getBatchSize()));
+      if (source.shardPreference != null) {
+        params.put("preference", "_shards:" + source.shardPreference);
+      }
+      HttpEntity queryEntity = new NStringEntity(query, ContentType.APPLICATION_JSON);
+      response =
+          restClient.performRequest("GET", endPoint, params, queryEntity, new BasicHeader("", ""));
+      JsonObject searchResult = parseResponse(response);
+      updateScrollId(searchResult);
+      batch = new ArrayList<>();
+      return updateBatchAndReadFirstDocument(searchResult);
+    }
+
+    private void updateScrollId(JsonObject searchResult) {
+      scrollId = searchResult.getAsJsonPrimitive("_scroll_id").getAsString();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      if (batch.size() > 0){
+        current = batch.remove(0);
```

Comment:
This is inefficient - removing the first element in an `ArrayList` is O(N). Just use an `Iterator` instead of `List` as the member variable.